### PR TITLE
Add NgModules for material

### PR DESCRIPTION
**Pull request by jelbourn**
R: @hansl @kara 

This PR includes:
* Temporarily move to head Angular until RC5 is released
* Introducing modules for each component and remove all `directives: [...]`
* Introducing an "all" module (`MaterialModule`)
* Bootstrapping via modules
* Update tests to use modules
* Workaround for https://github.com/angular/angular/issues/10503
* Bumps the version to alpha.6-3 (which I published under `@experimental` yesterday)
* Update to new `Provider` syntax

### CODE REVIEW

File: src/core/core.ts [hunk 1]
```diff
@@ -12,7 +12,8 @@ export {
 export {
```
Comment (hansl) [inline src/core/core.ts]: Export a core module from here, or export all sub-modules through this file.

**Comment by hansl** [comment]
Globally the feel is good. Please address my previous comment.

File: src/core/core.ts (hunk 1 above)
Comment (jelbourn) [inline src/core/core.ts]: Exported individual modules and added an `NgCoreModule`

File: src/components/all/all.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  MdSlideToggleModule,
+  MdTabsModule,
+  MdToolbarModule,
+  MdTooltipModule,
+  OverlayModule,
+  PortalModule,
+  RtlModule,
+];
+
+@NgModule(({
```
Comment (soyzamudio) [inline src/components/all/all.ts]: one too many parenthesis here?

File: src/components/button-toggle/button-toggle.spec.ts [hunk 3]
```diff
@@ -488,9 +480,15 @@ class ButtonTogglesInsideButtonToggleGroupMultiple {
...
 
 @Component({
-  directives: [MD_BUTTON_TOGGLE_DIRECTIVES],
   template: `
   <md-button-toggle>Yes</md-button-toggle>
   `
 })
 class StandaloneButtonToggle { }
+
+const TEST_COMPONENTS = [
```
Comment (hansl) [inline src/components/button-toggle/button-toggle.spec.ts]: These are only used once, please inline them.

File: src/components/checkbox/checkbox.spec.ts [hunk 4]
```diff
@@ -583,30 +578,38 @@ class CheckboxWithTabIndex {
...
 @Component({
-  directives: [MdCheckbox],
   template: `<md-checkbox (change)="lastEvent = $event"></md-checkbox>`
 })
 class CheckboxWithChangeEvent {
   lastEvent: MdCheckboxChange;
 }
+
+
+const TEST_COMPONENTS = [
```
Comment (hansl) [inline src/components/checkbox/checkbox.spec.ts]: Same as above.

File: src/components/dialog/dialog-container.ts [hunk 5]
```diff
@@ -18,15 +18,14 @@ import {MdDialogContentAlreadyAttachedError} from './dialog-errors';
...
-  directives: [PortalHostDirective],
   host: {
     'class': 'md-dialog-container',
     '[attr.role]': 'dialogConfig?.role'
   }
 })
 export class MdDialogContainer extends BasePortalHost implements AfterViewInit {
   /** The portal host inside of this container into which the dialog content will be loaded. */
-  @ViewChild(PortalHostDirective) private _portalHost: PortalHostDirective;
+  @ViewChild(PortalHostDirective) _portalHost: PortalHostDirective;
```
Comment (hansl) [inline src/components/dialog/dialog-container.ts]: ?

File: src/components/dialog/dialog.spec.ts [hunk 6]
```diff
@@ -122,7 +124,7 @@ function detectChangesForDialogOpen(fixture: ComponentFixture<ComponentWithChild
   // Two rounds of change detection are necessary: one to *create* the dialog container, and
   // another to cause the lifecycle events of the container to run and load the dialog content.
   fixture.detectChanges();
-  setTimeout(() => fixture.detectChanges(), 50);
+  setTimeout(() => fixture.detectChanges(), 150);
```
Comment (hansl) [inline src/components/dialog/dialog.spec.ts]: Nit; is this necessary and can we make it async?  150 is really long for a unittest.

File: src/components/icon/icon.spec.ts [hunk 7]
```diff
@@ -437,9 +441,16 @@ class MdIconFromSvgUrlTestApp {
...
   selector: 'test-app',
   template: `<md-icon [svgIcon]="iconName" [aria-label]="ariaLabel"></md-icon>`,
-  directives: [MdIcon],
 })
 class MdIconFromSvgNameTestApp {
   ariaLabel: string = null;
   iconName = '';
 }
+
+const TEST_COMPONENTS = [
```
Comment (hansl) [inline src/components/icon/icon.spec.ts]: Ditto.

File: src/components/input/input.spec.ts [hunk 8]
```diff
@@ -810,16 +796,34 @@ class MdInputWithBlurAndFocusEvents {
...
   selector: 'test-input-controller',
   template: `
     <md-input name="some-name"></md-input>
-  `,
-  directives: [MdInput]
+  `
 })
 class MdInputWithNameTestController {}
+
+const TEST_COMPONENTS = [
```
Comment (hansl) [inline src/components/input/input.spec.ts]: Idem.

File: src/components/radio/radio.spec.ts [hunk 9]
```diff
@@ -538,3 +536,9 @@ function dispatchFocusChangeEvent(eventName: string, element: HTMLElement): void
   event.initEvent(eventName, true, true);
   element.dispatchEvent(event);
 }
+
+const TEST_COMPONENTS = [
```
Comment (hansl) [inline src/components/radio/radio.spec.ts]: gleich

File: src/components/sidenav/sidenav.spec.ts [hunk 10]
```diff
@@ -355,3 +358,10 @@ class BasicTestApp {
     this.closeCount++;
   }
 }
+
+
+const TEST_COMPONENTS = [
```
Comment (hansl) [inline src/components/sidenav/sidenav.spec.ts]: "

File: src/core/core.ts [hunk 11]
```diff
@@ -24,23 +33,46 @@ export {OverlayState} from './overlay/overlay-state';
...
 
 export {
   MdUniqueSelectionDispatcher,
   MdUniqueSelectionDispatcherListener
 } from './coordination/unique-selection-dispatcher';
+
+export {MdLineModule, MdLine, MdLineSetter} from './line/line';
+
+
+let coreModules = [
```
Comment (hansl) [inline src/core/core.ts]: const?

File: src/core/portal/portal.spec.ts [hunk 12]
```diff
@@ -13,18 +16,29 @@ import {
...
+      imports: [PortalModule],
+      declarations: TEST_COMPONENTS,
+    });
   }));
 
+  beforeEach(fakeAsync(inject([TestComponentBuilder], (tcb: TestComponentBuilder) => {
+    builder = tcb;
+  })));
+
+  // beforeEach(inject([TestComponentBuilder], fakeAsync((tcb: TestComponentBuilder) => {
```
Comment (hansl) [inline src/core/portal/portal.spec.ts]: Commented code.

File: src/core/core.ts (hunk 11 above)
Comment (jelbourn) [inline src/core/core.ts]: Done

File: src/core/portal/portal.spec.ts (hunk 12 above)
Comment (jelbourn) [inline src/core/portal/portal.spec.ts]: Leftover from debugging, removed.

File: src/components/dialog/dialog.spec.ts (hunk 6 above)
Comment (jelbourn) [inline src/components/dialog/dialog.spec.ts]: Leftover from debugging, reverted.
(and will likely go away when the zone fix is in)

**Comment by jelbourn** [comment]
Removed all `TEST_COMPONENTS`

**Comment by jelbourn** [comment]
@hansl all comments should be resolved

**Comment by hansl** [comment]
LGTM.

File: src/components/all/all.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+const MATERIAL_MODULES = [
+  MdButtonModule,
+  MdButtonToggleModule,
+  MdCardModule,
+  MdCheckboxModule,
+  MdDialogModule,
+  MdGridListModule,
+  MdIconModule,
+  MdInputModule,
```
Comment (LeoLun) [inline src/components/all/all.ts]: asdad

**Pull request merged**